File: skills/tokensmind-agent-network-runtime/scripts/python_runtime/system_credential_store.py
import json
import os
import shutil
import subprocess

from .portable_store import RECORD_NAMES, _validate_record

SERVICE_NAME = "TokensMind Agent Network"
APPLICATION_NAME = "tokensmind-agent-network"
MACOS_SECURITY_PATH = "/usr/bin/security"
MACOS_ITEM_NOT_FOUND = 44


def _validate_record_name(name):
    if name not in RECORD_NAMES:
        raise ValueError("Unsupported portable state record: %s" % name)


def _parse_record(name, source):
    try:
        value = json.loads(source)
    except json.JSONDecodeError as error:
        raise ValueError(
            "System Agent Network %s credential is invalid JSON" % name,
        ) from error
    _validate_record(name, value)
    return value


def _command_failure(backend, operation, result):
    return RuntimeError(
        "%s %s failed with exit code %s" % (backend, operation, result.returncode),
    )


def run_credential_command(args, input_text=""):
    return subprocess.run(
        args,
        input=input_text,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=False,
    )


class MacOSCredentialStore:
    def __init__(self, namespace, runner=run_credential_command):
        self._namespace = namespace
        self._runner = runner

    def _account(self, name):
        _validate_record_name(name)
        return "%s:%s" % (self._namespace, name)

    def read(self, name):
        result = self._runner([
            MACOS_SECURITY_PATH, "find-generic-password",
            "-a", self._account(name), "-s", SERVICE_NAME, "-w",
        ])
        if result.returncode == MACOS_ITEM_NOT_FOUND:
            return None
        if result.returncode != 0:
            raise _command_failure("macOS Keychain", "read", result)
        return _parse_record(name, result.stdout.strip())

    def write(self, name, value):
        _validate_record(name, value)
        result = self._runner([
            MACOS_SECURITY_PATH, "add-generic-password",
            "-a", self._account(name), "-s", SERVICE_NAME, "-U", "-w",
        ], json.dumps(value, separators=(",", ":")) + "\n")
        if result.returncode != 0:
            raise _command_failure("macOS Keychain", "write", result)

    def remove(self, name):
        result = self._runner([
            MACOS_SECURITY_PATH, "delete-generic-password",
            "-a", self._account(name), "-s", SERVICE_NAME,
        ])
        if result.returncode not in (0, MACOS_ITEM_NOT_FOUND):
            raise _command_failure("macOS Keychain", "remove", result)


class LinuxSecretServiceStore:
    def __init__(self, namespace, command, runner=run_credential_command):
        self._namespace = namespace
        self._command = command
        self._runner = runner

    def _attributes(self, name):
        _validate_record_name(name)
        return [
            "application", APPLICATION_NAME,
            "origin", self._namespace,
            "record", name,
        ]

    def read(self, name):
        result = self._runner([self._command, "lookup"] + self._attributes(name))
        missing = (
            result.returncode == 1
            and not result.stdout.strip()
            and not result.stderr.strip()
        )
        if missing:
            return None
        if result.returncode != 0:
            raise _command_failure("Linux Secret Service", "read", result)
        return _parse_record(name, result.stdout.strip())

    def write(self, name, value):
        _validate_record(name, value)
        result = self._runner(
            [self._command, "store", "--label=%s" % SERVICE_NAME]
            + self._attributes(name),
            json.dumps(value, separators=(",", ":")) + "\n",
        )
        if result.returncode != 0:
            raise _command_failure("Linux Secret Service", "write", result)

    def remove(self, name):
        if self.read(name) is None:
            return
        result = self._runner([self._command, "clear"] + self._attributes(name))
        if result.returncode != 0:
            raise _command_failure("Linux Secret Service", "remove", result)


def resolve_system_credential_store(
        platform, namespace, env=None, *, runner=run_credential_command):
    if platform == "darwin":
        if not os.path.isfile(MACOS_SECURITY_PATH):
            return None
        if not os.access(MACOS_SECURITY_PATH, os.X_OK):
            return None
        return MacOSCredentialStore(namespace, runner)
    if platform == "linux":
        environment = os.environ if env is None else env
        command = shutil.which("secret-tool", path=environment.get("PATH"))
        if command:
            return LinuxSecretServiceStore(namespace, command, runner)
    return None

File: skills/tokensmind-agent-network-runtime/scripts/lib/runtime/systemCredentialStore.js
import { spawn as defaultSpawn } from 'node:child_process';
import { constants as fsConstants, promises as defaultFs } from 'node:fs';
import path from 'node:path';
import { validatePortableRecord } from './portableStore.js';

const SERVICE_NAME = 'TokensMind Agent Network';
const APPLICATION_NAME = 'tokensmind-agent-network';
const RECORD_NAMES = new Set(['active', 'pending', 'instance']);
const MACOS_SECURITY_PATH = '/usr/bin/security';
const MACOS_ITEM_NOT_FOUND = 44;

function validateRecordName(name) {
  if (!RECORD_NAMES.has(name)) throw new Error(`Unsupported portable state record: ${name}`);
}

function credentialAccount(namespace, name) {
  validateRecordName(name);
  return `${namespace}:${name}`;
}

function parseRecord(name, source) {
  let value;
  try {
    value = JSON.parse(source);
  } catch (error) {
    throw new Error(`System Agent Network ${name} credential is invalid JSON`, { cause: error });
  }
  validatePortableRecord(name, value);
  return value;
}

function commandFailure(backend, operation, result) {
  return new Error(
    `${backend} ${operation} failed with exit code ${String(result.code)}`,
  );
}

export function createCommandRunner({ spawnImpl = defaultSpawn } = {}) {
  return ({ command, args, input = '' }) => new Promise((resolve, reject) => {
    const child = spawnImpl(command, args, { stdio: ['pipe', 'pipe', 'pipe'] });
    const stdout = [];
    const stderr = [];
    let settled = false;
    child.stdout.on('data', (chunk) => stdout.push(chunk));
    child.stderr.on('data', (chunk) => stderr.push(chunk));
    child.stdin.on('error', (error) => {
      if (error.code === 'EPIPE' || settled) return;
      settled = true;
      reject(error);
    });
    child.on('error', (error) => {
      if (settled) return;
      settled = true;
      reject(error);
    });
    child.on('close', (code) => {
      if (settled) return;
      resolve({
        code,
        stdout: Buffer.concat(stdout).toString('utf8'),
        stderr: Buffer.concat(stderr).toString('utf8'),
      });
    });
    child.stdin.end(input);
  });
}

function createMacOperations({ namespace, run }) {
  return {
    async read(name) {
      const account = credentialAccount(namespace, name);
      const result = await run({
        command: MACOS_SECURITY_PATH,
        args: ['find-generic-password', '-a', account, '-s', SERVICE_NAME, '-w'],
      });
      if (result.code === MACOS_ITEM_NOT_FOUND) return null;
      if (result.code !== 0) throw commandFailure('macOS Keychain', 'read', result);
      return parseRecord(name, result.stdout.trim());
    },
    async write(name, value) {
      validateRecordName(name);
      validatePortableRecord(name, value);
      const account = credentialAccount(namespace, name);
      const result = await run({
        command: MACOS_SECURITY_PATH,
        args: ['add-generic-password', '-a', account, '-s', SERVICE_NAME, '-U', '-w'],
        input: `${JSON.stringify(value)}\n`,
      });
      if (result.code !== 0) throw commandFailure('macOS Keychain', 'write', result);
    },
    async remove(name) {
      const account = credentialAccount(namespace, name);
      const result = await run({
        command: MACOS_SECURITY_PATH,
        args: ['delete-generic-password', '-a', account, '-s', SERVICE_NAME],
      });
      if (![0, MACOS_ITEM_NOT_FOUND].includes(result.code)) {
        throw commandFailure('macOS Keychain', 'remove', result);
      }
    },
  };
}

function linuxAttributes(namespace, name) {
  validateRecordName(name);
  return ['application', APPLICATION_NAME, 'origin', namespace, 'record', name];
}

function createLinuxOperations({ namespace, command, run }) {
  async function lookup(name) {
    const result = await run({
      command,
      args: ['lookup', ...linuxAttributes(namespace, name)],
    });
    const missing = result.code === 1 && !result.stdout.trim() && !result.stderr.trim();
    if (missing) return null;
    if (result.code !== 0) throw commandFailure('Linux Secret Service', 'read', result);
    return parseRecord(name, result.stdout.trim());
  }
  return {
    read: lookup,
    async write(name, value) {
      validatePortableRecord(name, value);
      const result = await run({
        command,
        args: ['store', `--label=${SERVICE_NAME}`, ...linuxAttributes(namespace, name)],
        input: `${JSON.stringify(value)}\n`,
      });
      if (result.code !== 0) throw commandFailure('Linux Secret Service', 'write', result);
    },
    async remove(name) {
      if (await lookup(name) === null) return;
      const result = await run({
        command,
        args: ['clear', ...linuxAttributes(namespace, name)],
      });
      if (result.code !== 0) throw commandFailure('Linux Secret Service', 'remove', result);
    },
  };
}

async function isExecutable(fs, filePath) {
  try {
    await fs.access(filePath, fsConstants.X_OK);
    return true;
  } catch {
    return false;
  }
}

async function findExecutable({ name, env, fs }) {
  const directories = String(env.PATH || '').split(path.delimiter).filter(Boolean);
  for (const directory of directories) {
    const candidate = path.join(directory, name);
    if (await isExecutable(fs, candidate)) return candidate;
  }
  return null;
}

export async function resolveSystemCredentialStore({
  platform,
  namespace,
  env = process.env,
  fs = defaultFs,
  run = createCommandRunner(),
} = {}) {
  if (platform === 'darwin') {
    if (!await isExecutable(fs, MACOS_SECURITY_PATH)) return null;
    return createMacOperations({ namespace, run });
  }
  if (platform === 'linux') {
    const command = await findExecutable({ name: 'secret-tool', env, fs });
    return command ? createLinuxOperations({ namespace, command, run }) : null;
  }
  return null;
}
